iggtools/common/utils.py

```diff
@@ -271,5 +296,134 @@ def parse_table(lines, columns, decode=True):
         yield [values[i] for i in column_indexes]
 
 
+def retry(operation, MAX_TRIES=3):
+    # Note the use of a separate random generator for retries so transient
+    # errors won't perturb other random streams used in the application.
+    invocation = [0]
+    @wraps(operation)
+    def wrapped_operation(*args, **kwargs):
+        randgen = None
+        remaining_attempts = MAX_TRIES
+        delay = 2.0
+        while remaining_attempts > 1:
+            try:
+                t_start = time.time()
+                return operation(*args, **kwargs)
+            except:
+                t_end = time.time()
+                if randgen == None:
+                    invocation[0] += 1
+                    randgen = random.Random(os.getpid() * 10000 + invocation[0]).random
+                if t_end - t_start > 30:
+                    # For longer operations, the delay should increase, so that the backoff will meaningfully reduce load on the failing service.
+                    delay = (t_end - t_start) * 0.2
+                wait_time = delay * (1.0 + 2.0 * randgen())
+                tsprint(f"Sleeping {wait_time} seconds before retry {MAX_TRIES - remaining_attempts + 1} of {operation} with {args}, {kwargs}.")
+                time.sleep(wait_time)
+                delay *= 3.0
+                remaining_attempts -= 1
+        # The last attempt is outside try/catch so caller can handle exception
+        return operation(*args, **kwargs)
+    return wrapped_operation
+
+
+def suppress_exceptions(func):
+    @wraps(func)
+    def func_noexc(*args, **kwargs):
+        try:
+            return func(*args, **kwargs)
+        except:
+            tsprint(f"\nSuppressing exception below and returning None from {func} called with arguments {args} {kwargs}:\n{traceback.format_exc(limit=-1)}")
+            return None
+    return func_noexc
+
+
+def split(iterable, portion_size):
+    assert portion_size > 0
+    p = []
+    for item in iterable:
+        p.append(item)
+        if len(p) == portion_size:
+            yield p
+            p = []
+    if p:
+        yield p
+
+
+# The next few functions present a uniform interface to a parallel version of map
+# using either multiple threads (preferred) or multiple processes (necessary when
+# the function's performance is bottlenecked on CPU-intensive python code).
+#
+# Use multiprocessing *only* if the Python code itself is slow and bottlenecked on CPU.
+# If the problem is just I/O latency, use multithreading, rather than multiprocessing.
+# If the CPU intensive work can be done in subcommands, use multithreading rather than
+# multiprocessing, to invoke multiple subcommands in parallel.
+#
+# When multiprocessing, it pays little to exceed the number of physical cores,
+# and never more than 2x, particularly if there are merging costs.
+#
+# With multithreading, on the other hand, there is no reason to constrain
+# the number of threads;  by default we set it equal to the number of items.
+
```

Comment:
cc @zhaoc1 this may be interesting for you :)

Reply:
totally~~ thanks!